Sources/Types.swift

```diff
@@ -34,15 +34,73 @@ public typealias GetState = () -> State
 /// Typealias for the `Store.anyDispatch` function with the ability of managing the output with a promise
 public typealias AnyDispatch = (_: Dispatchable) -> Promise<Any>
 
-/// Entity capable of executing task asynchronously. This can be useful in tests to control asynchronous tasks
-public protocol AsyncProvider {
-  /// Calling this method will enqueue a closure to be executed asynchronously
+/// See `Executor`
+@available(*, deprecated, message: "Use the new `Executor` instead")
+public typealias AsyncProvider = Executor
+
+/// Entity capable of executing a closure. This can be useful in tests to control where and when tasks
+/// are executed.
+///
+/// - warning: Katana assumes that implementations of this protocol will always guarantee that the tasks
+/// given as arguments are never executed in parallel.
+public protocol Executor {
+  /// Calling this method will enqueue a task to be executed asynchronously
+  @available(*, deprecated, message: "Use the new `executeAsync` instead")
   func execute(_ closure: @escaping () -> Void)
+
+  /// Calling this method will immediately execute a task, blocking the current thread until the task is
+  /// finished. It needs to be reentrant (it can be called again from the closure argument without
+  /// causing a deadlock).
+  @discardableResult
+  func executeSync<T>(_ closure: @escaping () -> T) -> T
 }
 
-/// DispatchQueue conformance to `AsyncProvider` so that `DispatchQueue.main` can be used
-extension DispatchQueue: AsyncProvider {
+extension Executor {
+  /// Calling this method will enqueue a task to be executed asynchronously
+  ///
+  /// Note: this is the method that supersedes `execute(_:)`
+  public func executeAsync(_ closure: @escaping () -> Void) {
+    self.execute(closure)
+  }
+
+  /// Calling this method will immediately execute a task, blocking the current thread until the task is
+  /// finished
+  ///
+  /// This is an inefficient default implementation made to avoid having a breaking change in the code. It
+  /// will be removed in the future.
+  @available(*, deprecated, message: "Use a proper primitive given by your underlying implementation")
+  public func executeSync<T>(_ closure: @escaping () -> T) -> T {
+    let promise = Promise<T>()
+
+    self.executeAsync {
+      promise.resolve(closure())
+    }
+
+    return try! Hydra.await(promise) // swiftlint:disable:this force_try
+  }
+}
```

Comment:
This is blocking the thread, so it is not reentrant (i.e., if called from the thread where `executeAsync` would be called it would deadlock)